src/compat/db_json.cpp

```diff
@@ -0,0 +1,941 @@
+ * doc (in): every possible inner doc
+ * current_path (in)
+ * search_str (in)
+ * one_or_all (in): 0 is one and 1 and all; one returns only the first occurence
+ */
+
+static void
+db_json_search_helper (const JSON_VALUE *whole_doc,
+                       const JSON_VALUE *doc,
+                       const char *current_path,
+                       const char *search_str,
+                       int one_or_all,
+                       std::vector < std::string > &result)
+{
+  if (one_or_all == 0 && result.size () == 1)
+    {
+      return;
+    }
+
+  if (!doc->IsArray () && !doc->IsObject ())
+    {
+      JSON_POINTER p (current_path);
+      JSON_VALUE *resulting_json;
+
+      if (p.IsValid () && (resulting_json = p.Get (*const_cast<JSON_VALUE *> (whole_doc))) != NULL)
+        {
+          char final_string[DB_JSON_MAX_STRING_SIZE];
+
+          if (resulting_json->IsInt ())
+            {
+              snprintf (final_string, DB_JSON_MAX_STRING_SIZE, "%d", resulting_json->GetInt ());
+            }
+          else if (resulting_json->IsDouble ())
+            {
+              snprintf (final_string, DB_JSON_MAX_STRING_SIZE, "%lf", resulting_json->GetDouble ());
+            }
+          else if (resulting_json->IsString ())
+            {
+              strncpy (final_string, resulting_json->GetString (), DB_JSON_MAX_STRING_SIZE);
+            }
+
+          if (strstr (final_string, search_str) != NULL)
+            {
+              result.push_back (std::string (current_path));
+            }
+        }
+      else
+        {
+          //everything should be valid
+          assert (false);
+        }
+    }
+
+  if (doc->IsArray ())
+    {
+      int index = 0;
+      for (JSON_VALUE::ConstValueIterator itr = doc->Begin (); itr != doc->End (); ++itr)
+        {
+          char index_str[3];
+          snprintf (index_str, 2, "%d", index);
+
+          char *next_path = (char *) db_private_alloc (NULL, strlen (current_path) + strlen (index_str) + 2);
+          strcpy (next_path, current_path);
+          strcat (next_path, "/");
+          strcat (next_path, index_str);
+
+          db_json_search_helper (whole_doc, itr, next_path, search_str, one_or_all, result);
+
+          index++;
+          db_private_free (NULL, next_path);
+        }
+    }
+  else if (doc->IsObject ())
+    {
+      for (JSON_VALUE::ConstMemberIterator itr = doc->MemberBegin (); itr != doc->MemberEnd (); ++itr)
+        {
+          char *next_path =
+            (char *) db_private_alloc (NULL, strlen (current_path) + 1 + strlen (itr->name.GetString ()) + 1);
+          strcpy (next_path, current_path);
+          strcat (next_path, "/");
+          strcat (next_path, itr->name.GetString ());
+
+          db_json_search_helper (whole_doc, &itr->value, next_path, search_str, one_or_all, result);
+
+          db_private_free (NULL, next_path);
+        }
+    }
+}
+
+void
+db_json_add_member_to_object (JSON_DOC *doc, char *name, char *value)
```

Comment:
1. Should name and value be const char *?
2. Please explain in a code comment how stringref works and why it is used here.